### PR TITLE
add the parseEscapeExpression function to fix singleID issue.

### PR DESCRIPTION
**Pull request by neraliu**

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: src/handlebars-utils.js [hunk 1]
```diff
@@ -177,6 +182,26 @@ HandlebarsUtils.isValidExpression = function(input, i, type) {
     return re;
 };
 
+// @function HandlebarsUtils.parseEscapeExpression
+HandlebarsUtils.parseEscapeExpression = function(str) {
+  var j=0, inSquareBracket = false,
+      fstr = '', re = [];
+
+  str = str.trim(); // the new regexp will capture the tailing space.
```
Comment (adon-at-work) [inline src/handlebars-utils.js]: String.trim() is unfriendly to IE 8 or below
https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/trim

File: src/handlebars-utils.js [hunk 2]
```diff
@@ -35,32 +35,32 @@ HandlebarsUtils.rawEndBlockRegExp = /^\{\{\{\{\/([^\s!"#%&'\(\)\*\+,\.\/;<=>@\[\
 /* '{{{' '~'? 'space'* '@'? 'space'* ('not {}~'+) 'space'* ('not {}~'+) '~'? non-greedy '}}}' and not follow by '}' */
 HandlebarsUtils.RAW_EXPRESSION = 1; // {{{expression}}}
 // HandlebarsUtils.rawExpressionRegExp = /^\{\{\{\s*([^\s!"#%&'\(\)\*\+,\.\/;<=>\[\\\]\^`\{\|\}\~]+)\s*?\}\}\}(?!})/;
-HandlebarsUtils.rawExpressionRegExp = /^\{\{\{~?\s*@?\s*([^\s\}\{~]+)\s*([^\}\{~]*)~??\}\}\}(?!})/;
+HandlebarsUtils.rawExpressionRegExp = /^\{\{\{~?\s*@?\s*([^\s\}\{~]+)\s*([^\}\{~]*)~?\}\}\}(?!})/;
 
 /* '{{' '~'? 'space'* '@'? 'space'* ('not {}~'+) 'space'* ('not {}~'+) '~'? non-greedy '}}' and not follow by '}' */
```
Comment (adon-at-work) [inline src/handlebars-utils.js]: can you please either remove or update this comment according to the new regexp?

File: src/handlebars-utils.js [hunk 3]
```diff
@@ -122,6 +122,11 @@ HandlebarsUtils.isValidExpression = function(input, i, type) {
         case HandlebarsUtils.ESCAPE_EXPRESSION:
             re = HandlebarsUtils.escapeExpressionRegExp.exec(s);
             if (re !== null && re[1] !== undefined) {
```
Comment (adon-at-work) [inline src/handlebars-utils.js]: minor change suggestion: `if (re) {` is sufficient

File: src/handlebars-utils.js [hunk 4]
```diff
@@ -177,6 +182,26 @@ HandlebarsUtils.isValidExpression = function(input, i, type) {
...
+HandlebarsUtils.parseEscapeExpression = function(str) {
+  var j=0, inSquareBracket = false,
+      fstr = '', re = [];
+
+  str = str.trim(); // the new regexp will capture the tailing space.
+  while(j<str.length) {
+      // space is defined as \s in the handlebars lex
+      if (str[j].match(/\s/) && !inSquareBracket) break;
+      if (str[j] === '[') inSquareBracket = true;
+      if (str[j] === ']') inSquareBracket = false;
```
Comment (adon-at-work) [inline src/handlebars-utils.js]: use `else if`

**Comment by neraliu** [comment]
all feedback has been addressed in this check in
https://github.com/yahoo/secure-handlebars/commit/5dac988445f833f624ca43a7beffbd7c3dda002a

**Comment by adon-at-work** [comment]
look gd. thanks

**Pull request merged**